File: sdk/python/hitechcloud/resources/cloud_instance.py
"""Cloud Instance resource (25 endpoints)"""

from .base import BaseResource

class CloudInstanceResource(BaseResource):
    """Cloud Instance: full VM lifecycle management"""

    def list_vms(self, service_id: int) -> dict:
        """GET /api/service/{id}/vms - List virtual servers"""
        return self._http.get(f"/api/service/{service_id}/vms")

    def create_vm(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms - Create new virtual server"""
        return self._http.post(f"/api/service/{service_id}/vms", kwargs)

    def get_vm(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid} - Get VM details"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}")

    def destroy_vm(self, service_id: int, vmid: int) -> dict:
        """DELETE /api/service/{id}/vms/{vmid} - Remove virtual server"""
        return self._http.delete(f"/api/service/{service_id}/vms/{vmid}")

    def resize_vm(self, service_id: int, vmid: int, **kwargs) -> dict:
        """PUT /api/service/{id}/vms/{vmid} - Edit/resize virtual server"""
        return self._http.put(f"/api/service/{service_id}/vms/{vmid}", kwargs)

    def stop_vm(self, service_id: int, vmid: int) -> dict:
        """POST /api/service/{id}/vms/{vmid}/stop - Stop virtual server"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/stop")

    def start_vm(self, service_id: int, vmid: int) -> dict:
        """POST /api/service/{id}/vms/{vmid}/start - Start virtual server"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/start")

    def reboot_vm(self, service_id: int, vmid: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms/{vmid}/reboot - Reboot virtual server"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/reboot", kwargs)

    def reset_password(self, service_id: int, vmid: int) -> dict:
        """POST /api/service/{id}/vms/{vmid}/resetpwd - Reset root password"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/resetpwd")

    def rebuild_vm(self, service_id: int, vmid: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms/{vmid}/rebuild - Rebuild server"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/rebuild", kwargs)

    def list_vm_ips(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/ips - List IPs assigned to VM"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/ips")

    def assign_ip(self, service_id: int, vmid: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms/{vmid}/ips - Assign IP to VM"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/ips", kwargs)

    def list_interfaces(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/interfaces - List network interfaces"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/interfaces")

    def get_cpu_usage(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/usage/cpu - CPU usage graph"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/usage/cpu")

    def get_network_usage(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/usage/net - Network usage graph"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/usage/net")

    def get_disk_usage(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/usage/disk - Disk usage graph"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/usage/disk")

    def list_disks(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/storage - List disks attached to VM"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/storage")

    def resize_disk(self, service_id: int, vmid: int, diskid: int, **kwargs) -> dict:
        """PUT /api/service/{id}/vms/{vmid}/storage/{diskid} - Resize disk"""
        return self._http.put(f"/api/service/{service_id}/vms/{vmid}/storage/{diskid}", kwargs)

    def list_iso_images(self, service_id: int) -> dict:
        """GET /api/service/{id}/images - List ISO images"""
        return self._http.get(f"/api/service/{service_id}/images")

    def add_iso_image(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/images - Add ISO image"""
        return self._http.post(f"/api/service/{service_id}/images", kwargs)

    def list_available_ips(self, service_id: int, vmid: int, iface: str) -> dict:
        """GET /api/service/{id}/vms/{vmid}/interfaces/{iface}/ips - List available IPs"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/interfaces/{iface}/ips")

    def remove_ip(self, service_id: int, vmid: int, ipid: int) -> dict:
        """DELETE /api/service/{id}/vms/{vmid}/ips/{ipid} - Remove IP from VM"""
        return self._http.delete(f"/api/service/{service_id}/vms/{vmid}/ips/{ipid}")

    def get_rdns(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/rdns - Get reverse DNS entries"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/rdns")

    def update_rdns(self, service_id: int, vmid: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms/{vmid}/rdns - Update reverse DNS entries"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/rdns", kwargs)

    def rebuild_network(self, service_id: int, vmid: int) -> dict:
        """POST /api/service/{id}/vms/{vmid}/rebuild_network - Rebuild network"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/rebuild_network")

File: sdk/ruby/lib/hitechcloud/resources/cloud_instance.rb
module HiTechCloud
  module Resources
    class CloudInstance < Base

        # List VMs
        def vms(id)
          @http.get("/api/service/%{id}/vms" % { id: id })
        end

        # Create VM
        def create_vm(id, data)
          @http.post("/api/service/%{id}/vms" % { id: id }, data)
        end

        # Get VM
        def vm(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}" % { id: id, vmid: vmid })
        end

        # Delete VM
        def delete_vm(id, vmid)
          @http.delete("/api/service/%{id}/vms/%{vmid}" % { id: id, vmid: vmid })
        end

        # Update VM
        def update_vm(id, vmid, data)
          @http.put("/api/service/%{id}/vms/%{vmid}" % { id: id, vmid: vmid }, data)
        end

        # Stop VM
        def stop_vm(id, vmid)
          @http.post("/api/service/%{id}/vms/%{vmid}/stop" % { id: id, vmid: vmid })
        end

        # Start VM
        def start_vm(id, vmid)
          @http.post("/api/service/%{id}/vms/%{vmid}/start" % { id: id, vmid: vmid })
        end

        # Reboot VM
        def reboot_vm(id, vmid)
          @http.post("/api/service/%{id}/vms/%{vmid}/reboot" % { id: id, vmid: vmid })
        end

        # Reset password
        def reset_password(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/resetpwd" % { id: id, vmid: vmid }, data)
        end

        # Rebuild VM
        def rebuild_vm(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/rebuild" % { id: id, vmid: vmid }, data)
        end

        # Get VM IPs
        def vm_ips(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/ips" % { id: id, vmid: vmid })
        end

        # Add VM IP
        def add_vm_ip(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/ips" % { id: id, vmid: vmid }, data)
        end

        # Get VM interfaces
        def vm_interfaces(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/interfaces" % { id: id, vmid: vmid })
        end

        # Get CPU usage
        def cpu_usage(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/usage/cpu" % { id: id, vmid: vmid })
        end

        # Get network usage
        def net_usage(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/usage/net" % { id: id, vmid: vmid })
        end

        # Get disk usage
        def disk_usage(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/usage/disk" % { id: id, vmid: vmid })
        end

        # Get storage
        def storage(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/storage" % { id: id, vmid: vmid })
        end

        # Update storage
        def update_storage(id, vmid, diskid, data)
          @http.put("/api/service/%{id}/vms/%{vmid}/storage/%{diskid}" % { id: id, vmid: vmid, diskid: diskid }, data)
        end

        # List images
        def images(id)
          @http.get("/api/service/%{id}/images" % { id: id })
        end

        # Create image
        def create_image(id, data)
          @http.post("/api/service/%{id}/images" % { id: id }, data)
        end

        # Get interface IPs
        def interface_ips(id, vmid, iface)
          @http.get("/api/service/%{id}/vms/%{vmid}/interfaces/%{iface}/ips" % { id: id, vmid: vmid, iface: iface })
        end

        # Delete VM IP
        def delete_vm_ip(id, vmid, ipid)
          @http.delete("/api/service/%{id}/vms/%{vmid}/ips/%{ipid}" % { id: id, vmid: vmid, ipid: ipid })
        end

        # Get VM rDNS
        def vm_rdns(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/rdns" % { id: id, vmid: vmid })
        end

        # Set VM rDNS
        def set_vm_rdns(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/rdns" % { id: id, vmid: vmid }, data)
        end

        # Rebuild network
        def rebuild_network(id, vmid)
          @http.post("/api/service/%{id}/vms/%{vmid}/rebuild_network" % { id: id, vmid: vmid })
      end
    end
  end
end
